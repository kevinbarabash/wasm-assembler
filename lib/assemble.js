import {encodeUInt32} from './leb128'
import {sectionTypes, externalTypes} from './constants'
import instructions from './instructions'

// TODO: handle other types
const assembleFuncType = (type) => {
    return [
        0x60,   // types.func
        ...encodeUInt32(type.params.length),
        ...type.params,
        ...encodeUInt32(type.result.length),
        ...type.result,
    ]
}

const assembleCodeBody = (body) => [...encodeUInt32(body.length), ...body]

const assembleExport = (exp) => {
    return [
        ...encodeUInt32(exp.name.length),
        ...exp.name.split('').map(c => c.charCodeAt(0)),
        externalTypes[exp.type],
        ...encodeUInt32(exp.index),
    ]
}

const assembleImport = (exp) => {
    if (exp.type === 'func') {
        return [
            ...encodeUInt32(exp.name1.length),
            ...exp.name1.split('').map(c => c.charCodeAt(0)),
            ...encodeUInt32(exp.name2.length),
            ...exp.name2.split('').map(c => c.charCodeAt(0)),
            externalTypes[exp.type],
            ...encodeUInt32(exp.index),
        ]
    } else if (exp.type === 'memory') {
        return [
            ...encodeUInt32(exp.name1.length),
            ...exp.name1.split('').map(c => c.charCodeAt(0)),
            ...encodeUInt32(exp.name2.length),
            ...exp.name2.split('').map(c => c.charCodeAt(0)),
            externalTypes[exp.type],
            0,   // limits: flags
            ...encodeUInt32(exp.min),
        ]
    } else {

    }
}

const assembleTypeIndex = (funcIndex) => [...encodeUInt32(funcIndex)]

const assembleData = (exp) => {
    return [
        ...encodeUInt32(exp.index),
        ...exp.offset,
        instructions.end,
        ...encodeUInt32(exp.data.length),
        ...exp.data,
    ]
}

const assembleSection = (section, type) => {
    if (type === 'start') {
        const name = section[0].name;
        const funcIndex = encodeUInt32(section[0].index)
        return [
            sectionTypes.start,
            ...encodeUInt32(funcIndex.length),
            ...funcIndex,
        ]
    }

    const assembleChild = {
        'functype': assembleFuncType,
        'typeidx': assembleTypeIndex,
        'code': assembleCodeBody,
        'export': assembleExport,
        'import': assembleImport,
        'data': assembleData,
    }[type]

    const concatedChildren = section.reduce(
        (accum, child) => accum.concat(assembleChild(child)), [])

    const numberOfChildren = encodeUInt32(section.length)

    return [
        sectionTypes[type],
        ...encodeUInt32(numberOfChildren.length + concatedChildren.length),
        ...numberOfChildren,
        ...concatedChildren,
    ]
}

const assembleProgram = (program) => {
    const concatedSections = Object.entries(program).reduce(
        (accum, [type, section]) =>
            accum.concat(assembleSection(section, type)), [])

    const bytes = new Uint8Array([
        0x00, 0x61, 0x73, 0x6D,     // magic number
        0x01, 0x00, 0x00, 0x00,     // version = 1, little endian
        ...concatedSections,
    ])

    var buffer = new Buffer(bytes.length);
    for (var i = 0; i < bytes.length; i++) {
        buffer[i] = bytes[i];
    }

    return buffer
}

module.exports = assembleProgram
