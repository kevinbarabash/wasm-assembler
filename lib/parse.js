const assert = require('assert');
const parse = require('s-expression');

const {types, i32, get_local, end} = require('./constants.js');
const instructions = require('./instructions.js');
const {encodeUInt32} = require('./leb128.js');

const program = {
    functype: [],
    import: [],
    typeidx: [],
    export: [],
    code: [],
};

const funcIdx = [];

const processModule = (module) => {
    const [type, ...children] = module;
    assert.equal(type, 'module');

    children.forEach(process);
}

const processFunc = (func) => {
    const funcName = func[1]
    const params = func.filter(child => Array.isArray(child) && child[0] === 'param')
    const results = func.filter(child => Array.isArray(child) && child[0] === 'result')
    const locals = func.filter(child => Array.isArray(child) && child[0] === 'local')
    assert(results.length < 2, "can return at most one result")

    const localIdx = [
        ...params.map(param => param.length === 3 ? param[1] : ""),
        ...locals.map(local => local.length === 3 ? local[1] : ""),
    ]

    const signature = {
        params: params.map(param => types[param[param.length - 1]]),
        result: results.map(result => types[result[1]]),
    }

    const insts = func.slice(2 + params.length + results.length + locals.length);
    const code = [
        ...encodeUInt32(locals.length), // local decl count
    ];

    if (locals.length > 0) {
        code.push(...encodeUInt32(locals.length)) // local type count
        code.push(...locals.map(local => types[local[local.length - 1]]))
    }

    let i = 0;
    while (i < insts.length) {
        const inst = insts[i++];
        if (inst === 'i32.const') {
            code.push(instructions[inst])
            code.push(...encodeUInt32(insts[i++]))
        } else if (inst === 'get_local') {
            code.push(instructions[inst])
            const local = insts[i++]
            if (local[0] === '$') {
                code.push(localIdx.indexOf(local))
            } else {
                code.push(...encodeUInt32(local))
            }
        } else if (inst === 'set_local') {
            code.push(instructions[inst])
            const local = insts[i++]
            if (local[0] === '$') {
                code.push(...encodeUInt32(localIdx.indexOf(local)))
            } else {
                code.push(...encodeUInt32(local))
            }
        } else if (inst === 'call') {
            code.push(instructions[inst])
            const func = insts[i++]
            if (func[0] === '$') {
                code.push(...encodeUInt32(funcIdx.indexOf(func)))
            } else {
                code.push(...encodeUInt32(func))
            }
        } else if (inst === 'if') {
            code.push(instructions[inst])
            // skip result type if one exists
            if (['i32', 'i64', 'f32', 'f64'].includes(insts[i])) {
                code.push(types[insts[i]]);
                i++;
            }
        } else {
            code.push(instructions[inst])
        }
    }
    code.push(instructions.end)

    // TODO: figure out how to reuse types
    // Right now we take advantage of the fact that each function has its own
    // type.
    program.typeidx.push(program.functype.length)
    program.functype.push(signature)
    program.code.push(code)
    funcIdx.push(funcName)
}

const processImport = (exp) => {
    const func = exp[3]
    const funcName = func[1]

    const params = func.filter(child => Array.isArray(child) && child[0] === 'param')
    const results = func.filter(child => Array.isArray(child) && child[0] === 'result')

    const signature = {
        params: params.map(param => types[param[param.length - 1]]),
        result: results.map(result => types[result[1]]),
    }

    program.functype.push(signature)
    funcIdx.push(funcName)
    program.import.push({
        name1: exp[1],
        name2: exp[2],
        type: 'func',   // TODO: change these to 'kind'
        index: funcIdx.indexOf(funcName),
    })
}

const processExport = (exp) => {
    program.export.push({
        name: exp[1],
        type: exp[2][0],
        index: funcIdx.indexOf(exp[2][1]),
    })
}

const process = (expr) => {
    switch (expr[0]) {
        case 'module':
            processModule(expr);
            break;
        case 'func':
            processFunc(expr);
            break;
        case 'import':
            processImport(expr);
            break;
        case 'export':
            processExport(expr);
            break;
        default:
            throw new Error(`can't process '${expr[0]}' expressions yet`);
    }

    return program;
}

module.exports = (code) => {
    const codeWithoutComments = code.replace(/;;.*/g, '')
    return process(parse(codeWithoutComments))
}
