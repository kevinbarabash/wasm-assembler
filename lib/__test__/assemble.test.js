import fs from 'fs'
import path from 'path'

import parse from '../parse'
import assemble from '../assemble'

const success_files = fs.readdirSync(path.join(__dirname, 'success_files'));
const error_files = fs.readdirSync(path.join(__dirname, 'error_files'))

describe('assemble', () => {
    describe('success', () => {
        success_files.filter(name => /\.wast$/.test(name))
            .map(inName =>
                test(`${inName}`, () => {

                    const code = fs.readFileSync(
                        path.join(__dirname, 'success_files', inName)
                    ).toString()
                    const output = assemble(parse(code))

                    const outName = inName.replace(/\.wast$/, '.wasm')
                    const expected = fs.readFileSync(
                        path.join(__dirname, 'success_files', outName)
                    )

                    expect(output).toEqual(expected)
                }
        ))
    })

    describe('error handling', () => {
        error_files.map(filename =>
            test(`${filename.replace('.wast', '')}`, () => {
                expect(() => {
                    const code = fs.readFileSync(
                        path.join(__dirname, 'error_files', filename)
                    ).toString()

                    assemble(parse(code))
                }).toThrowErrorMatchingSnapshot()
            })
        )

        // TODO: check for number functions that are out of bounds of funcIdx
        // TODO: add tests for missing types, arguments, etc.
    })
})
