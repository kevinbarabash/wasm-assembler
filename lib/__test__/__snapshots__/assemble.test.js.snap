// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assemble error handling export-undefined-function 1`] = `"export: function $add not defined"`;

exports[`assemble error handling func-collision 1`] = `"func $f1 already defined"`;

exports[`assemble error handling invalid-instruction 1`] = `"(return i32) instruction not recognized"`;

exports[`assemble error handling start-undefined-function 1`] = `"start: function $add not defined"`;
